src/openrct2-ui/windows/MapGen.cpp

```diff
@@ -374,1013 +295,1093 @@ constexpr int32_t WATERLEVEL_MIN = 0;
-            WindowMapgenChangeMapSize(-1);
-            w->Invalidate();
-            break;
-        case WIDX_BASE_HEIGHT_UP:
-            _baseHeight = std::min(_baseHeight + 2, BASESIZE_MAX);
-            w->Invalidate();
-            break;
-        case WIDX_BASE_HEIGHT_DOWN:
-            _baseHeight = std::max(_baseHeight - 2, BASESIZE_MIN);
-            w->Invalidate();
-            break;
-        case WIDX_WATER_LEVEL_UP:
-            _waterLevel = std::min(_waterLevel + 2, WATERLEVEL_MAX);
-            w->Invalidate();
-            break;
-        case WIDX_WATER_LEVEL_DOWN:
-            _waterLevel = std::max(_waterLevel - 2, WATERLEVEL_MIN);
-            w->Invalidate();
-            break;
-        case WIDX_FLOOR_TEXTURE:
-            LandTool::ShowSurfaceStyleDropdown(w, widget, _floorTexture);
-            break;
-        case WIDX_WALL_TEXTURE:
-            LandTool::ShowEdgeStyleDropdown(w, widget, _wallTexture);
-            break;
+        int32_t type;
+
+        switch (widgetIndex)
+        {
+            case WIDX_FLOOR_TEXTURE:
+                if (dropdownIndex == -1)
+                    dropdownIndex = gDropdownHighlightedIndex;
+
+                type = (dropdownIndex == -1) ? _floorTexture : dropdownIndex;
+
+                if (gLandToolTerrainSurface == type)
+                {
+                    gLandToolTerrainSurface = OBJECT_ENTRY_INDEX_NULL;
+                }
+                else
+                {
+                    gLandToolTerrainSurface = type;
+                    _floorTexture = type;
+                }
+                Invalidate();
+                break;
+            case WIDX_WALL_TEXTURE:
+                if (dropdownIndex == -1)
+                    dropdownIndex = gDropdownHighlightedIndex;
+
+                type = (dropdownIndex == -1) ? _wallTexture : dropdownIndex;
+
+                if (gLandToolTerrainEdge == type)
+                {
+                    gLandToolTerrainEdge = OBJECT_ENTRY_INDEX_NULL;
+                }
+                else
+                {
+                    gLandToolTerrainEdge = type;
+                    _wallTexture = type;
+                }
+                Invalidate();
+                break;
+        }
     }
-}
 
-static void WindowMapgenBaseDropdown(WindowBase* w, WidgetIndex widgetIndex, int32_t dropdownIndex)
-{
-    int32_t type;
+    void BaseUpdate()
+    {
+        // Tab animation
+        if (++frame_no >= TabAnimationLoops[page])
+            frame_no = 0;
+        InvalidateWidget(WIDX_TAB_1);
+    }
 
-    switch (widgetIndex)
+    void TextInput(WidgetIndex widgetIndex, std::string_view text)
     {
-        case WIDX_FLOOR_TEXTURE:
-            if (dropdownIndex == -1)
-                dropdownIndex = gDropdownHighlightedIndex;
+        int32_t value;
+        char* end;
 
-            type = (dropdownIndex == -1) ? _floorTexture : dropdownIndex;
+        value = strtol(u8string(text).c_str(), &end, 10);
```

Comment:
end will be an invalid pointer due to the u8string(text) temporary. You'll have to create a string from the text as a separate line.
```cpp
const auto strText = u8string(text);
value = strtol(strText.c_str(), &end, 10);
```